opencensus/test_exporter.go

```diff
@@ -0,0 +1,98 @@
+package opencensus
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"go.opencensus.io/metric/metricdata"
+	"go.opencensus.io/metric/metricexport"
+	"go.opencensus.io/stats/view"
+)
+
+// TestExporter keeps exported metric data in memory to aid in testing the instrumentation.
+//
+// Metrics can be retrieved with `GetPoint()`. In order to deterministically retrieve the most recent values, you must first invoke `ReadAndExport()`.
+type TestExporter struct {
+	// points is a map from a label signature to the latest value for the time series represented by the signature.
+	// Use function `labelSignature` to get a signature from a `metricdata.Metric`.
+	points       map[string]metricdata.Point
+	metricReader *metricexport.Reader
+}
+
+// NewTestExporter returns a new exporter.
+func NewTestExporter(metricReader *metricexport.Reader) *TestExporter {
+	return &TestExporter{points: make(map[string]metricdata.Point), metricReader: metricReader}
+}
+
+// ExportMetrics records the view data.
+func (e *TestExporter) ExportMetrics(ctx context.Context, data []*metricdata.Metric) error {
+	for _, metric := range data {
+		for _, ts := range metric.TimeSeries {
+			signature := labelSignature(metric.Descriptor.Name, labelObjectsToKeyValue(metric.Descriptor.LabelKeys, ts.LabelValues))
+			e.points[signature] = ts.Points[len(ts.Points)-1]
+		}
+	}
+	return nil
+}
+
+// GetPoint returns the latest point for the time series identified by the given labels.
+func (e *TestExporter) GetPoint(metricName string, labels map[string]string) (metricdata.Point, bool) {
+	v, ok := e.points[labelSignature(metricName, labelMapToKeyValue(labels))]
+	return v, ok
+}
+
+// ReadAndExport reads the current values for all metrics and makes them available to this exporter.
+func (e *TestExporter) ReadAndExport() {
+	// The next line forces the view worker to process all stats.Record* calls that
+	// happened within Store() before the call to ReadAndExport below. This abuses the
+	// worker implementation to work around lack of synchronization.
+	// TODO(jkohen,rghetia): figure out a clean way to make this deterministic.
+	view.SetReportingPeriod(time.Minute)
```

Comment:
@rghetia any thoughts on what we could do here? This is currently used in client_test.go in this same PR, which deterministically fails without this (or a sleep).

I understand that the cause is that the worker queues record commands [in this function](https://github.com/census-instrumentation/opencensus-go/blob/master/stats/view/worker.go#L109-L117) and there's no way to wait for them to be processed. I know a Flush command is a cheap way out. Even a Stop would work for testing.

Reply:
view.Start() and view.Stop() would be better. 10 millisecond sleep should work and won't be noticeable in the test.

Reply:
Thanks. view.Start and view.Stop don't seem to exist. Is it worth adding them? I believe SetReportingPeriod should work reliably, as long as the worker implementation doesn't change. Any plans to change it?

Reply:
There is no plan to change SetReportingPeriod. It does work but it is not clean as you pointed out earlier.

Reply:
At, got it. Let's either revisit this with OpenTelemetry or address in a separate PR if we move this file to OpenCensus.